src/internal/_curry3.js

```diff
@@ -12,43 +13,51 @@ var _curry2 = require('./_curry2');
-      return _curry1(function(c) { return fn(a, b, c); });
-    } else if (n === 3 && a != null && a['@@functional/placeholder'] === true &&
-                          b != null && b['@@functional/placeholder'] === true &&
-                          c != null && c['@@functional/placeholder'] === true) {
-      return f3;
-    } else if (n === 3 && a != null && a['@@functional/placeholder'] === true &&
-                          b != null && b['@@functional/placeholder'] === true) {
-      return _curry2(function(a, b) { return fn(a, b, c); });
-    } else if (n === 3 && a != null && a['@@functional/placeholder'] === true &&
-                          c != null && c['@@functional/placeholder'] === true) {
-      return _curry2(function(a, c) { return fn(a, b, c); });
-    } else if (n === 3 && b != null && b['@@functional/placeholder'] === true &&
-                          c != null && c['@@functional/placeholder'] === true) {
-      return _curry2(function(b, c) { return fn(a, b, c); });
-    } else if (n === 3 && a != null && a['@@functional/placeholder'] === true) {
-      return _curry1(function(a) { return fn(a, b, c); });
-    } else if (n === 3 && b != null && b['@@functional/placeholder'] === true) {
-      return _curry1(function(b) { return fn(a, b, c); });
-    } else if (n === 3 && c != null && c['@@functional/placeholder'] === true) {
-      return _curry1(function(c) { return fn(a, b, c); });
-    } else {
-      return fn(a, b, c);
+    switch (arguments.length) {
+      case 0:
+        return f3;
+
+      case 1:
+        if (_isPlaceholder(a)) {
+          return f3;
+        }
+        return _curry2(function(_b, _c) { return fn(a, _b, _c); });
+
+      case 2:
+        if (_isPlaceholder(a)) {
+          if (_isPlaceholder(b)) {
+            return f3;
+          } else {
+            return _curry2(function(_a, _c) { return fn(_a, b, _c); });
+          }
+        } else if (_isPlaceholder(b)) {
+          return _curry2(function(_b, _c) { return fn(a, _b, _c); });
+        }
+        return _curry1(function(_c) { return fn(a, b, _c); });
+
+      default:
+        if (_isPlaceholder(a)) {
+          if (_isPlaceholder(b)) {
+            if (_isPlaceholder(c)) {
+              return f3;
+            } else {
+              return _curry2(function(_a, _b) { return fn(_a, _b, c); });
+            }
+          } else if (_isPlaceholder(c)) {
+            return _curry2(function(_a, _c) { return fn(_a, b, _c); });
+          } else {
+            return _curry1(function(_a) { return fn(_a, b, c); });
+          }
+        } else if (_isPlaceholder(b)) {
+          if (_isPlaceholder(c)) {
+            return _curry2(function(_b, _c) { return fn(a, _b, _c); });
+          } else {
+            return _curry1(function(_b) { return fn(a, _b, c); });
+          }
+        } else if (_isPlaceholder(c)) {
+          return _curry1(function(_c) { return fn(a, b, _c); });
+        }
+        return fn(a, b, c);
```

Comment:
I do find this more difficult to follow than the existing implementation (which itself is pretty scary).